### ISSUE TITLE
Negative Energy Cost

### ISSUE THREAD

**Issue by biemann** [fzi-forschungszentrum-informatik/tropical_precooling_environment#1]
Hi, 

I have seen your presentation at last BuildSys conference and I share your viewpoints. I am trying to make an RL algorithm that is able to learn in this environment and see if my tested results generalise to this environment. I turned in the following issue.

The Energy cost, described in Equation 6 of your paper can be potentially negative when the setpoint temperature is higher than the zone temperature. Now this is not supposed to happen with a trained algorithm, but happens at the beginning of the training.  The problem is when we want to minimise energy cost, the agent will choose actions that lead to negative energy cost, which is not the desired behaviour at all.

I am just wondering how you would approach this problem. I am thinking about the following ideas:

- Turn cooling off, when Tz < T_SP. 
- Write E(t)=|Q'(t)|e(t)/COP.
- Give a negative reward when Tz < T_SP

I think this can also be a problem when calculating the performance measure in Equation 7

**Comment by david-woelfle** [comment]
Hi @biemann 

thank you for opening this issue, that is a good point you are addressing. I think the option of simulating the AC to turn off once the setpoint exceeds the current zone temperature is probably the most realistic. I would also like to ensure that there are no discontinuous step changes resulting from this switch operation. AFAIK, the AC system would probably switch off the supply air fans at some point if the room is too cold already. Hence, to keep things as realistic as possible I would simulate this by changing eq. (5) such that:

![image](https://user-images.githubusercontent.com/29716719/110613115-2a9b6480-8191-11eb-9c7d-64ca19a475a9.png)

What do you think?

**Comment by biemann** [comment]
Hi,

Thanks for the answer. I agree that is probably the best way to address the problem. It is important to avoid discontinuities. I did some experiments and it seems to almost always give positive energy costs. There is still possibly an issue when the supply air temperature is higher than the zone temperature. If I give realistic values to the setpoint temperature, this did never occur.

This leads me to two other questions:

- Is the supply air temperature always the same than for the measured data, or is it simulated as well?
- Is there a reasonable range for setpoint temperatures I should give the agent? Maybe I am missing something, as you mention this in the paper in Section 4.7

**Comment by david-woelfle** [comment]
Hi,

sorry for the late reply. I was pretty busy with project work these days and your issue here slipped a bit under the table. Anyway, I have now finished implementing a fixed version and it should no longer be possible to exploit the negative energy cost issue you found. The updated code lives [in the V2 branch here](https://github.com/fzi-forschungszentrum-informatik/tropical_precooling_environment/tree/V2), and I would be happy if you could check the new version operates as you would expect. I would then merge the branch to master shortly after.

Regarding your other points:
- Yes the supply air temperature is a measured quantity and not simulated.
- That is a good point about the possible action space. I have also updated the code here, and you now set to arbitrary float values (including NaN representing AC off) as actions. However, these are now internally clipped to a range of 10...35°C. Please see the code and docstring of the `env.step()` method for details.

**Comment by biemann** [comment]
Hi, 

Sorry for the late reply as well. I was also kind of busy. When I run some experiments, I implemented an equivalent code, so I think it works as expected.

Otherwise a few other general remarks I had when testing out this environment (not related to this issue):

- I normalised the state space, as a change of 1 degree in temperature has a significantly larger impact than having a change of 1 $ in energy cost.
- I also normalised the action space, as this is required by many RL algorithms (SAC, PPO). Now I understand that these changes are specific to RL and not necessary for MPC applications (I think), so I am not sure if it would be of interest to implement these changes to the environment.
- Otherwise I found it quite challenging to train an algorithm in this environment, probably because of the huge action space. I reduced the dimension of the action space to a lower dimension and repeat the same action during a few hours.
- But even then, it is still not easy, because the state space is still very large and has temporal dependencies between the states. I was trying to use some neural networks that are able to handle such dependencies. I think a good solution is to use a (1D)-CNN as a feature extractor or use some type of dimensionality reduction technique that is able to handle temporal dependencies (such as Kalman filters) before passing it to an MLP.
- As far as I am aware of, studying a case study like this one, where we receive a batch of observations and then have to plan a multi-step strategy before receiving the next batch, is not studied at all in the RL literature.

I am quite interested in trying to make something work in this environment, but I will probably try to study the use of RNNs/CNNs in an easier case study first.

**Comment by david-woelfle** [comment]
Hi,

thank you once more for your feedback! IMHO the normalization is pretty standard in ML, and I would see it as a part of the ML algorithm engineering process. I thus think it's fine that the environment returns unnormalized data, as a real building would likely do the same.

Also interesting to hear about how challenging the env was for you. As I designed it I rather thought this to be a simple problem. Besides, from a controls point of view, I still think it is quite simple though. However, it seems to be a repeating pattern in RL, that even simple problems (like the classic controls tasks in OpenAI gym) require a surprising high amount of attention and engineering before being solved with RL. Whatever we might take away from that insight ;)

**Comment by david-woelfle** [comment]
V2 branch has now been merged, which resolves this issue.
